File: passenger/src/test/java/com/modsen/passenger/service/impl/PassengerServiceImplTest.java
package com.modsen.passenger.service.impl;

import com.modsen.passenger.dto.response.PassengerResponse;
import com.modsen.passenger.exception.EmailAlreadyExistsException;
import com.modsen.passenger.exception.PassengerNotFoundException;
import com.modsen.passenger.exception.PhoneNumberAlreadyExistsException;
import com.modsen.passenger.mapper.PassengerMapper;
import com.modsen.passenger.model.Passenger;
import com.modsen.passenger.repository.PassengerRepository;
import com.modsen.passenger.repository.RatingRepository;
import com.modsen.passenger.util.TestData;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@ExtendWith(MockitoExtension.class)
class PassengerServiceImplTest {

    @Mock
    private PassengerMapper passengerMapper;

    @Mock
    private RatingRepository ratingRepository;

    @Mock
    private PassengerRepository passengerRepository;

    @InjectMocks
    private PassengerServiceImpl passengerService;

    @Test
    void findById_shouldThrowPassengerNotFoundException() {
        doReturn(Optional.empty())
                .when(passengerRepository)
                .findById(TestData.PASSENGER_ID);

        assertThrows(PassengerNotFoundException.class,
                () -> passengerService.findById(TestData.PASSENGER_ID));

        verify(passengerRepository).findById(eq(TestData.PASSENGER_ID));
    }

    @Test
    void findById_shouldReturnExpectedResponse() {
        PassengerResponse expected = TestData.defaultPassengerResponse();
        Passenger passenger = TestData.defaultPassenger();

        doReturn(Optional.of(passenger))
                .when(passengerRepository)
                .findById(TestData.PASSENGER_ID);
        doReturn(expected)
                .when(passengerMapper)
                .toResponse(passenger);

        PassengerResponse actual = passengerService.findById(TestData.PASSENGER_ID);

        assertThat(actual).isEqualTo(expected);

        verify(passengerRepository).findById(eq(TestData.PASSENGER_ID));
        verify(passengerMapper).toResponse(eq(passenger));
    }

    @Test
    void findAll_shouldReturnExpectedPage() {
        List<Passenger> content = List.of(
                TestData.defaultPassenger(),
                TestData.defaultPassenger()
        );
        Pageable pageable = Pageable.ofSize(2);

        doReturn(new PageImpl<>(content))
                .when(passengerRepository)
                .findAll(pageable);
        doReturn(TestData.defaultPassengerResponse())
                .when(passengerMapper)
                .toResponse(TestData.defaultPassenger());

        Page<PassengerResponse> actual = passengerService.findAll(pageable);

        assertThat(actual).hasSize(2);
        assertThat(actual).contains(TestData.defaultPassengerResponse());

        verify(passengerRepository).findAll(eq(pageable));
        verify(passengerMapper, times(2)).toResponse(TestData.defaultPassenger());
    }

    @Test
    void update_shouldThrowPassengerNotFoundException() {
        doReturn(Optional.empty())
                .when(passengerRepository)
                .findById(TestData.PASSENGER_ID);

        assertThrows(PassengerNotFoundException.class,
                () -> passengerService.update(TestData.PASSENGER_ID, null));

        verify(passengerRepository).findById(eq(TestData.PASSENGER_ID));
    }

    @Test
    void update_shouldThrowEmailAlreadyExistsException() {
        doReturn(Optional.of(TestData.defaultPassenger()))
                .when(passengerRepository)
                .findById(TestData.PASSENGER_ID);
        doReturn(Optional.of(TestData.defaultPassenger()))
                .when(passengerRepository)
                .findByEmail(TestData.EMAIL);

        assertThrows(EmailAlreadyExistsException.class,
                () -> passengerService.update(TestData.PASSENGER_ID, TestData.defaultPassengerUpdate()));

        verify(passengerRepository).findById(eq(TestData.PASSENGER_ID));
        verify(passengerRepository).findByEmail(eq(TestData.EMAIL));
    }

    @Test
    void update_shouldThrowPhoneNumberAlreadyExistsException() {
        doReturn(Optional.of(TestData.defaultPassenger()))
                .when(passengerRepository)
                .findById(TestData.PASSENGER_ID);
        doReturn(Optional.of(TestData.defaultPassenger()))
                .when(passengerRepository)
                .findByPhoneNumber(TestData.PHONE_NUMBER);

        assertThrows(PhoneNumberAlreadyExistsException.class,
                () -> passengerService.update(TestData.PASSENGER_ID, TestData.defaultPassengerUpdate()));

        verify(passengerRepository).findById(eq(TestData.PASSENGER_ID));
        verify(passengerRepository).findByPhoneNumber(eq(TestData.PHONE_NUMBER));
    }

    @Test
    void update_shouldSaveUpdatedPassenger() {
        doReturn(Optional.of(TestData.defaultPassenger()))
                .when(passengerRepository)
                .findById(TestData.PASSENGER_ID);
        doReturn(Optional.empty())
                .when(passengerRepository)
                .findByPhoneNumber(TestData.PHONE_NUMBER);
        doReturn(Optional.empty())
                .when(passengerRepository)
                .findByEmail(TestData.EMAIL);

        passengerService.update(TestData.PASSENGER_ID, TestData.defaultPassengerUpdate());

        verify(passengerRepository).findById(eq(TestData.PASSENGER_ID));
        verify(passengerRepository).findByPhoneNumber(eq(TestData.PHONE_NUMBER));
        verify(passengerRepository).findByEmail(eq(TestData.EMAIL));
        verify(passengerMapper).mapIfNotNull(eq(TestData.defaultPassengerUpdate()), eq(TestData.defaultPassenger()));
        verify(passengerRepository).save(eq(TestData.defaultPassenger()));
    }

    @Test
    void deleteById_shouldCallDeleteMethod() {
        doNothing()
                .when(passengerRepository)
                .deleteById(TestData.PASSENGER_ID);

        passengerService.deleteById(TestData.PASSENGER_ID);

        verify(passengerRepository).deleteById(eq(TestData.PASSENGER_ID));
    }

    @Test
    void create_shouldReturnCreatedObject() {
        doReturn(TestData.defaultPassenger())
                .when(passengerMapper)
                .toModel(TestData.defaultPassengerCreate());
        doReturn(Optional.empty())
                .when(passengerRepository)
                .findByEmail(TestData.EMAIL);
        doReturn(Optional.empty())
                .when(passengerRepository)
                .findByPhoneNumber(TestData.PHONE_NUMBER);
        doReturn(TestData.defaultPassengerResponse())
                .when(passengerMapper)
                .toResponse(TestData.defaultPassenger());

        PassengerResponse actual = passengerService.create(TestData.defaultPassengerCreate());

        assertThat(actual).isEqualTo(TestData.defaultPassengerResponse());

        verify(passengerMapper).toModel(eq(TestData.defaultPassengerCreate()));
        verify(passengerMapper).toResponse(eq(TestData.defaultPassenger()));
        verify(passengerRepository).save(TestData.defaultPassenger());
        verify(ratingRepository).save(eq(TestData.initialRating()));
    }

    @Test
    void create_shouldThrowEmailAlreadyExistsException() {
        doReturn(Optional.of(TestData.defaultPassenger()))
                .when(passengerRepository)
                .findByEmail(TestData.EMAIL);

        assertThrows(EmailAlreadyExistsException.class,
                () -> passengerService.create(TestData.defaultPassengerCreate()));

        verify(passengerRepository).findByEmail(eq(TestData.EMAIL));
    }

    @Test
    void create_shouldThrowPhoneNumberAlreadyExistsException() {
        doReturn(Optional.of(TestData.defaultPassenger()))
                .when(passengerRepository)
                .findByPhoneNumber(TestData.PHONE_NUMBER);

        assertThrows(PhoneNumberAlreadyExistsException.class,
                () -> passengerService.create(TestData.defaultPassengerCreate()));

        verify(passengerRepository).findByPhoneNumber(eq(TestData.PHONE_NUMBER));
    }
}

File: passenger/src/test/kotlin/com/modsen/ktpassenger/service/impl/PassengerServiceImplTest.kt
package com.modsen.ktpassenger.service.impl

import com.modsen.ktpassenger.dto.PassengerResponse
import com.modsen.ktpassenger.dto.PassengerUpdate
import com.modsen.ktpassenger.exception.EmailAlreadyExistsException
import com.modsen.ktpassenger.exception.PassengerNotFoundException
import com.modsen.ktpassenger.exception.PhoneNumberAlreadyExistsException
import com.modsen.ktpassenger.model.Passenger
import com.modsen.ktpassenger.repository.PassengerRepository
import com.modsen.ktpassenger.util.TestData
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import java.util.*


@ExtendWith(MockitoExtension::class)
class PassengerServiceImplTest {
    @MockK
    lateinit var passengerRepository: PassengerRepository

    @InjectMockKs
    lateinit var passengerService: PassengerServiceImpl

    @BeforeEach
    fun setUp() = MockKAnnotations.init(this)

    @Test
    fun findById_shouldThrowPassengerNotFoundException() {
        every { passengerRepository.findById(TestData.PASSENGER_ID) } returns Optional.empty()

        assertThrows(PassengerNotFoundException::class.java) { passengerService.findById(TestData.PASSENGER_ID) }

        io.mockk.verify { passengerRepository.findById(TestData.PASSENGER_ID) }
    }

    @Test
    fun findById_shouldReturnExpectedResponse() {
        val expected: PassengerResponse = TestData.defaultPassengerResponse()
        val passenger: Passenger = TestData.defaultPassenger()
        every { passengerRepository.findById(TestData.PASSENGER_ID) } returns Optional.of(passenger)

        val actual: PassengerResponse = passengerService.findById(TestData.PASSENGER_ID)

        assertThat(actual).isEqualTo(expected)
        io.mockk.verify { passengerRepository.findById(TestData.PASSENGER_ID) }
    }

    @Test
    fun findAll_shouldReturnExpectedPage() {
        val content: List<Passenger> = listOf(TestData.defaultPassenger(), TestData.defaultPassenger())
        val pageable = Pageable.ofSize(2)
        every { passengerRepository.findAll(pageable) } returns PageImpl(content)

        val actual: Page<PassengerResponse> = passengerService.findAll(pageable)

        assertThat(actual).hasSize(2)
        assertThat(actual).contains(TestData.defaultPassengerResponse())
        io.mockk.verify { passengerRepository.findAll(pageable) }
    }

    @Test
    fun update_shouldThrowPassengerNotFoundException() {
        val request = PassengerUpdate(null, null, null, null)
        val stub: Optional<Passenger> = Optional.empty()
        every { passengerRepository.findById(TestData.PASSENGER_ID) } returns stub

        assertThrows<PassengerNotFoundException> {
            passengerService.update(TestData.PASSENGER_ID, request)
        }

        io.mockk.verify { passengerRepository.findById(TestData.PASSENGER_ID) }
    }

    @Test
    fun update_shouldThrowEmailAlreadyExistsException() {
        val passenger: Passenger = TestData.defaultPassenger()
        every { passengerRepository.findById(TestData.PASSENGER_ID) } returns Optional.of(passenger)
        every { passengerRepository.findByEmail(TestData.EMAIL) } returns Optional.of(passenger)

        assertThrows<EmailAlreadyExistsException> {
            passengerService.update(TestData.PASSENGER_ID, TestData.defaultPassengerUpdate())
        }

        io.mockk.verify { passengerRepository.findById(eq(TestData.PASSENGER_ID)) }
        io.mockk.verify { passengerRepository.findByEmail(eq(TestData.EMAIL)) }
    }

    @Test
    fun update_shouldThrowPhoneNumberAlreadyExistsException() {
        val passenger: Passenger = TestData.defaultPassenger()
        every { passengerRepository.findById(TestData.PASSENGER_ID) } returns Optional.of(passenger)
        every { passengerRepository.findByEmail(TestData.EMAIL) } returns Optional.empty()
        every { passengerRepository.findByPhoneNumber(TestData.PHONE_NUMBER) } returns Optional.of(passenger)

        assertThrows<PhoneNumberAlreadyExistsException> {
            passengerService.update(
                TestData.PASSENGER_ID,
                TestData.defaultPassengerUpdate()
            )
        }

        io.mockk.verify { passengerRepository.findById(TestData.PASSENGER_ID) }
        io.mockk.verify { passengerRepository.findByPhoneNumber(TestData.PHONE_NUMBER) }
    }

    @Test
    fun update_shouldSaveUpdatedPassenger() {
        val passenger: Passenger = TestData.defaultPassenger()
        every { passengerRepository.findById(TestData.PASSENGER_ID) } returns Optional.of(passenger)
        every { passengerRepository.findByPhoneNumber(TestData.PHONE_NUMBER) } returns Optional.empty()
        every { passengerRepository.findByEmail(TestData.EMAIL) } returns Optional.empty()
        every { passengerRepository.save(passenger) } returns passenger

        passengerService.update(TestData.PASSENGER_ID, TestData.defaultPassengerUpdate())

        io.mockk.verify { passengerRepository.findById(TestData.PASSENGER_ID) }
        io.mockk.verify { passengerRepository.findByPhoneNumber(TestData.PHONE_NUMBER) }
        io.mockk.verify { passengerRepository.findByEmail(TestData.EMAIL) }
        io.mockk.verify { passengerRepository.save(passenger) }
    }

    @Test
    fun deleteById_shouldCallDeleteMethod() {
        every { passengerRepository.deleteById(TestData.PASSENGER_ID) } returns Unit

        passengerService.deleteById(TestData.PASSENGER_ID)

        io.mockk.verify { passengerRepository.deleteById(TestData.PASSENGER_ID) }
    }

    @Test
    fun create_shouldReturnCreatedObject() {
        val passenger = TestData.defaultPassenger()
        val expected = TestData.defaultPassengerResponse()

        every { passengerRepository.findByEmail(TestData.EMAIL) } returns Optional.empty()
        every { passengerRepository.findByPhoneNumber(TestData.PHONE_NUMBER) } returns Optional.empty()
        every { passengerRepository.save(passenger.copy(id = null)) } answers {
            val savedPassenger = arg<Passenger>(0)
            savedPassenger.id = TestData.PASSENGER_ID
            savedPassenger
        }

        val actual = passengerService.create(TestData.defaultPassengerCreate())

        assertThat(actual).isEqualTo(expected)
        io.mockk.verify { passengerRepository.save(passenger) }
    }

    @Test
    fun create_shouldThrowEmailAlreadyExistsException() {
        val passenger: Passenger = TestData.defaultPassenger()
        every { passengerRepository.findByEmail(TestData.PHONE_NUMBER) } returns Optional.empty()
        every { passengerRepository.findByEmail(TestData.EMAIL) } returns Optional.of(passenger)

        assertThrows<EmailAlreadyExistsException> {
            passengerService.create(TestData.defaultPassengerCreate())
        }

        io.mockk.verify { passengerRepository.findByEmail(TestData.EMAIL) }
    }

    @Test
    fun create_shouldThrowPhoneNumberAlreadyExistsException() {
        val passenger: Passenger = TestData.defaultPassenger()
        every { passengerRepository.findByEmail(TestData.EMAIL) } returns Optional.empty()
        every { passengerRepository.findByPhoneNumber(TestData.PHONE_NUMBER) } returns Optional.of(passenger)

        assertThrows<PhoneNumberAlreadyExistsException> {
            passengerService.create(TestData.defaultPassengerCreate())
        }

        io.mockk.verify { passengerRepository.findByPhoneNumber(eq(TestData.PHONE_NUMBER)) }
    }
}
